### PR TITLE
Add sale defpact; add sale and transfer policy

### PR DESCRIPTION
**Pull request by sirlensalot**

### CODE REVIEW

File: pact/hft.pact [hunk 1]
```diff
@@ -369,6 +394,127 @@
...
+        (with-capability (BUY token seller buyer amount timeout (pact-id))
+          (buy token seller buyer buyer-guard amount))))
+  )
+
+  (defun offer
+    ( token:string
+      seller:string
+      amount:decimal
+    )
+    @doc "Initiate sale with by SELLER by escrowing AMOUNT of TOKEN until TIMEOUT."
```
Comment (sirlensalot) [inline pact/hft.pact:475]:
```suggestion
    @doc "Initiate sale with by SELLER by escrowing AMOUNT of TOKEN."
```

File: pact/hft.pact [hunk 2]
```diff
@@ -369,6 +394,127 @@
...
+    (debit token seller amount)
+    (credit token (sale-account) (create-pact-guard "SALE") amount)
+  )
+
+  (defun withdraw
+    ( token:string
+      seller:string
+      amount:decimal
+    )
+    @doc "Withdraw offer by SELLER of AMOUNT of TOKEN before TIMEOUT"
```
Comment (sirlensalot) [inline pact/hft.pact:486]:
```suggestion
    @doc "Withdraw offer by SELLER of AMOUNT of TOKEN "
```

File: pact/hft.pact [hunk 3]
```diff
@@ -369,6 +394,127 @@
...
+      seller:string
+      buyer:string
+      buyer-guard:guard
+      amount:decimal
+    )
+    @doc "Complete sale with transfer."
+    (require-capability (SALE_PRIVATE (pact-id)))
+    (debit token (sale-account) amount)
+    (credit token buyer buyer-guard amount)
+    (emit-event (TRANSFER token seller buyer amount))
```
Comment (sirlensalot) [inline pact/hft.pact:504]: Should probably emit transfer for all steps.

File: pact/fixed-quote-policy.pact [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    manifest:object{token-manifest.manifest}
+    (bind (get-policy token)
+      { 'mint-guard:=mint-guard:guard
+      , 'max-supply:=max-supply:decimal
+      }
+      (bind token )
+      (enforce-guard (at 'mint-guard (get-policy token)))
+      (enforce (<= (+ (at 'supply token))))
+    (enforce ())
+  )
```
Comment (ggobugi27) [inline pact/fixed-quote-policy.pact:69]:
```suggestion
  (defun enforce-mint:bool
    ( token:object{token-info}
      account:string
      amount:decimal
    )
    (bind (get-policy token)
      { 'mint-guard:=mint-guard:guard
      , 'max-supply:=max-supply:decimal
      }
      (enforce-guard (at 'mint-guard (get-policy token)))
      (enforce (<= (+ amount (at 'supply token)) max-supply))
      )
  )
 ```

File: pact/hft.pact [hunk 5]
```diff
@@ -369,6 +394,136 @@
...
+
+  (defpact sale
+    ( token:string
+      seller:string
+      amount:decimal
+      timeout:integer
+    )
+    (step-with-rollback
+      (with-capability (SALE token seller amount timeout (pact-id))
+        (offer token seller amount timeout))
```
Comment (ggobugi27) [inline pact/hft.pact:468]: `offer` function has no argument for `timeout`

File: pact/hft.pact [hunk 6]
```diff
@@ -369,6 +394,136 @@
...
+  )
+
+  (defun offer
+    ( token:string
+      seller:string
+      amount:decimal
+    )
+    @doc "Initiate sale with by SELLER by escrowing AMOUNT of TOKEN until TIMEOUT."
+    (require-capability (SALE_PRIVATE (pact-id)))
+    (debit token seller amount)
```
Comment (ggobugi27) [inline pact/hft.pact:486]: Looks like the capability `DEBIT` is not granted to this function. I also tried adding `(with-capability (OFFER ..))` to wrap around `debit` and `credit` but still doesn't resolve.

**Review by ggobugi27 — CHANGES_REQUESTED** [review]

**Review by ggobugi27 — APPROVED** [review]
Looks good to me.

**Pull request merged**